crates/binjs_io/src/context/huffman/mod.rs

```diff
@@ -0,0 +1,602 @@
+
+    pub fn from_bit_sequence(sequence: BitSequence) -> Self {
+        Self::new(sequence.bits, sequence.bit_len)
+    }
+
+    /// The bits in this Key.
+    ///
+    /// # Invariant
+    ///
+    /// Only the `self.bit_len()` lowest-weight bits may be non-0.
+    pub fn bits(&self) -> u32 {
+        self.0.bits
+    }
+
+    /// The number of bits of `bits` to use.
+    pub fn bit_len(&self) -> BitLen {
+        self.0.bit_len
+    }
+
+    pub fn as_bit_sequence(&self) -> &BitSequence {
+        &self.0
+    }
+
+    pub fn with_prefix(&self, prefix: &BitSequence) -> Self {
+        let sequence = self.0.with_prefix(prefix);
+        Key::from_bit_sequence(sequence)
+    }
+}
+
+/// A node in the Huffman tree.
+struct Node<T> {
+    /// The total number of instances of all `NodeContent::Leaf(T)` in this subtree.
+    instances: Instances,
+
+    /// The content of the node.
+    content: NodeContent<T>,
+}
+
+/// Contents of a node in the Huffman tree.
+enum NodeContent<T> {
+    /// A value from the stream of values.
+    Leaf(T),
+
+    /// An internal node obtained by joining two subtrees.
+    Internal {
+        left: Box<NodeContent<T>>,
+        right: Box<NodeContent<T>>,
+    },
+}
+
+/// Custom ordering of `NodeContent`.
+///
+/// We compare *only* by number of instances.
+impl<T> PartialOrd for Node<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self.instances.partial_cmp(&other.instances)
+    }
+}
+impl<T> Ord for Node<T> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.instances.cmp(&other.instances)
+    }
+}
+impl<T> PartialEq for Node<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.instances.eq(&other.instances)
+    }
+}
+impl<T> Eq for Node<T> {}
+
+/// Codebook associated to a sequence of values.
+#[derive(Clone, Debug)]
+pub struct Codebook<T> {
+    /// The longest bit length that actually appears in `mappings`.
+    highest_bit_len: BitLen,
+
+    /// The sequence of keys.
+    ///
+    /// Order is meaningful.
+    mappings: Vec<(T, Key)>,
+}
+
+impl<T> Codebook<T> {
+    /// The number of elements in this Codebook.
+    pub fn len(&self) -> usize {
+        self.mappings.len()
+    }
+
+    /// The longest bit length that acctually appears in this Codebook.
+    pub fn highest_bit_len(&self) -> BitLen {
+        self.highest_bit_len
+    }
+}
+
+impl<T> IntoIterator for Codebook<T> {
+    type Item = (T, Key);
+    type IntoIter = std::vec::IntoIter<(T, Key)>;
+    fn into_iter(self) -> Self::IntoIter {
+        self.mappings.into_iter()
+    }
+}
+
+impl<T> Codebook<T>
+where
+    T: Ord + Clone,
+{
+    /// Compute a `Codebook` from a sequence of values.
+    ///
+    /// Optionally, `max_bit_len` may specify a largest acceptable bit length.
+    /// If the `Codebook` may not be computed without exceeding this bit length,
+    /// fail with `Err(problemantic_bit_len)`.
+    ///
+    /// The current implementation only attempts to produce the best compression
```

Comment:
Put a TODO to address this.

You still need your error because someone could ask for an n-bit code for > 2^n items.

Reply:
I'll file an issue.